File: src/main/java/gargoyle/ct/prop/impl/simple/CTSimpleProperty.java
package gargoyle.ct.prop.impl.simple;

import gargoyle.ct.prop.impl.CTBaseObservableProperty;
import org.jetbrains.annotations.NotNull;

import java.text.MessageFormat;
import java.util.Objects;

public abstract class CTSimpleProperty<T> extends CTBaseObservableProperty<T> {
    private T value;

    @SuppressWarnings("unchecked")
    protected CTSimpleProperty(String name, T def) {
        super((Class<T>) def.getClass(), name);
        value = def;
    }

    protected CTSimpleProperty(Class<T> type, String name) {
        super(type, name);
    }

    @Override
    public void set(T value) {
        T oldValue = get();
        if (Objects.equals(oldValue, value)) {
            return;
        }
        this.value = value;
        Thread thread = firePropertyChange(value, oldValue);
        try {
            if (thread != null) {
                thread.join();
            }
        } catch (InterruptedException ex) {
            throw new IllegalStateException(ex);
        }
    }

    @Override
    public T get() {
        return value;
    }

    @NotNull
    @Override
    public String toString() {
        return MessageFormat.format("CTSimpleProperty'{'name=''{0}'', value={1}'}'", name, value);
    }
}

File: src/main/kotlin/gargoyle/ct/prop/impl/simple/CTSimpleProperty.kt
package gargoyle.ct.prop.impl.simple

import gargoyle.ct.prop.impl.CTBaseObservableProperty
import java.text.MessageFormat
import kotlin.reflect.KClass

abstract class CTSimpleProperty<T : Any> : CTBaseObservableProperty<T> {
    private lateinit var value: T

    @Suppress("UNCHECKED_CAST")
    protected constructor(name: String, def: T) : super(def::class as KClass<T>, name) {
        value = def
    }

    protected constructor(type: KClass<T>, name: String) : super(type, name)

    override fun set(value: T) {
        val oldValue = get()
        if (oldValue == value) {
            return
        }
        this.value = value
        val thread = firePropertyChange(value, oldValue)
        try {
            thread?.join()
        } catch (ex: InterruptedException) {
            throw IllegalStateException(ex)
        }
    }

    override fun get(): T {
        return value
    }

    override fun toString(): String {
        return MessageFormat.format("CTSimpleProperty'{'name=''{0}'', value={1}'}'", name, value)
    }
}
